Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py

```diff
@@ -0,0 +1,155 @@
+""" polls data from DS to azure logs """
+import logging
+from . import DS_api
+from . import AS_api
+from .state_serializer import State
+import json
+from . import constant
+
+logger = logging.getLogger("DS_poller")
+
+class poller:
+
+    def __init__(self, ds_id, ds_key, secret, as_id, as_key, connection_string, historical_days, url):
+        """ 
+            initializes all necessary variables from other classes for polling 
+        """
+        
+        self.DS_obj = DS_api.api(ds_id, ds_key, secret, url)
+        self.AS_obj = AS_api.logs_api(as_id, as_key)
+        self.date = State(connection_string)
+        logger.info("got inside the poller code")
+        self.event = self.date.get_last_event(historical_days)
+        if(isinstance(self.event, tuple)):
+            self.after_time = self.event[0]
+            self.before_time = self.event[1]
+            logger.info("From time: %s", self.after_time)
+            logger.info("to time: %s", self.before_time)
+        else:
+            logger.info("Polling from event number " + str(self.event))
+
+    def parse_desc(self, data):
+        """
+            adds more newlines to description for good display on Azure sentinel incidents
+        """
+        arr = data.splitlines()
+        res = ""
+        for e in arr:
+            res = res + e + "\n\n"
+        return res
+
+
+    def post_azure(self, response, item):
+        """
+            posts to azure after appending triage information on it
+        """
+        json_obj = response.json()
+        for it in item:
+            for i in range(len(json_obj)):
+                if(it['source']['alert-id'] == json_obj[i]['id'] or it['source']['incident-id'] == json_obj[i]['id']):
+
+                    comment_data = self.DS_obj.get_triage_comments(it['id'])
+                    json_obj[i]['status'] = it['state']
+                    json_obj[i]['triage_id'] = it['id']
+                    json_obj[i]['triage_raised_time'] = it['raised']
+                    json_obj[i]['triage_updated_time'] = it['updated']
+                    
+                    comment_data_filtered = []
+                    for comment in comment_data:
+                        comment['user-name'] = comment['user']['name']
+                        del comment['triage-item-id']
+                        del comment['updated']
+                        del comment['user']
+                        if comment['content'] != "":
+                            comment_data_filtered.append(comment)
```

Comment:
Similar comment to the above about copying rather than mutating. Also if you use my structure above:

```py
azure_obj = {
  ...alert_or_incident,
  status: triage_item['state'],
  triage_id: triage_item['id'],
  comments: [] # empty list here
}
```

then this part becomes:

```py
                    for comment in comment_data:
                        if not comment['content']: # may as well do this first
                          continue
                        azure_obj['comments'].append({
                          'user_name': comment['user']['name'],
                          'content': comment['content'],
                          'id': comment['id'],
                          'created': comment['created']
                        })
```

Another benefit of doing this is that if we add more field they aren't automatically dragged into the integration - which gives us a bit of control as to how they are introduced.

I also noticed that we are being inconsistent over our formatting - in the existing code we have `user-name` as a field whereas elsewhere we've used underscores, e.g. `triage_item_id`. Ideally we align them.